File: day4_part1_2019.dart

import 'dart:io';

void main() {
  List<String> input = File('input.txt').readAsLinesSync();
  List<int> range = input[0].split('-').map(int.parse).toList();

  int count = 0;
  for (int i = range[0]; i <= range[1]; i++) {
    String password = i.toString();
    if (hasAdjacentDigits(password) && neverDecrease(password)) {
      count++;
    }
  }

  print(count);
}

bool hasAdjacentDigits(String password) {
  for (int i = 0; i < password.length - 1; i++) {
    if (password[i] == password[i + 1]) {
      return true;
    }
  }
  return false;
}

bool neverDecrease(String password) {
  for (int i = 0; i < password.length - 1; i++) {
    if (int.parse(password[i]) > int.parse(password[i + 1])) {
      return false;
    }
  }
  return true;
}

File: day4_part1_2019.rs

use std::fs;

fn main() {
    let input = fs::read_to_string("input.txt").unwrap();
    let range: Vec<u32> = input.trim().split("-").map(|x| x.parse().unwrap()).collect();
    
    let mut count = 0;
    for num in range[0]..=range[1] {
        let digits: Vec<u32> = num.to_string().chars().map(|c| c.to_digit(10).unwrap()).collect();
        let mut has_adjacent = false;
        let mut never_decrease = true;
        
        for i in 1..digits.len() {
            if digits[i] == digits[i - 1] {
                has_adjacent = true;
            }
            if digits[i] < digits[i - 1] {
                never_decrease = false;
                break;
            }
        }
        
        if has_adjacent && never_decrease {
            count += 1;
        }
    }
    
    println!("{}", count);
}
